### ISSUE TITLE
WebAssembly support

### ISSUE THREAD

**Issue by Shnatsel** [rust-secure-code/cargo-auditable#53]
It is technically possible to support WebAssembly, since they do allow custom sections: https://webassembly.github.io/spec/core/appendix/custom.html

This may be useful since the overhead of the audit info is just a few kilobytes and WebAssembly is being applied not just for the web.

**Comment by Niederb** [comment]
I'm interested in this. Especially in combination with WebAssembly Components. Do you know if there was any progress? (from your side or from WebAssembly side)

**Comment by Shnatsel** [comment]
The WebAssembly spec seemingly allows this via custom sections. I haven't been working on this myself, but I am happy to accept a pull request adding support.

`cargo auditable` works by emitting an object for the linker to link into the final binary. The logic for writing the object file is largely just copy-pasted from rustc. Our copy is here: https://github.com/rust-secure-code/cargo-auditable/blob/24df0e1101e66050f22cb9a63cf9474c1526e34d/cargo-auditable/src/object_file.rs#L1

Since I last copied it, the compiler has added support for emitting WASM: https://github.com/rust-lang/rust/blob/03994e498df79aa1f97f7bbcfd52d57c8e865049/compiler/rustc_codegen_ssa/src/back/metadata.rs#L629-L671

So I think we'd just need to copy-paste the WASM emitting code from the compiler and that should be it, barring any linker shenanigans, which should also be possible to ape from the compiler.

**Comment by bjorn3** [comment]
> Since I last copied it, the compiler has added support for emitting WASM

That is for the `lib.rmeta` archive member which is supposed to be discarded by the linker. Emitting the `.rustc` section in dylibs (as done by `create_compressed_metadata_file`) is not supported by rustc on wasm yet.

**Comment by Niederb** [comment]
Thanks for the information and the fast response.
I hope to mentor a student/intern to work on this in half a year or so. So it will take a while, but hopefully we can create a PR then.

**Comment by brooksmtownsend** [comment]
Hey! I was really interested in your CTA, and we actually do something very similar for [wasmCloud](https://wasmcloud.com/) via the crate [wascap](https://crates.io/crates/wascap). We generate some information, slap it in a JWT, and then embed it in a custom section in the Wasm component and it works beautifully. Not saying we would use wascap here (we shouldn't) but I'm pretty interested in either doing or helping lift the logic there to apply to cargo-auditable

**Comment by Shnatsel** [comment]
> embed it in a custom section in the Wasm component

Could you point me to the code which does that? I need to see if it does that through passing arguments to the linker (as the rest of `cargo auditable` currently works) or by modifying the resulting WASM file, which I don't think could be easily adapted since `cargo auditable` does not know the location of the resulting WASM binary.

**Comment by brooksmtownsend** [comment]
You got it, will send a few links. That might be part of the issue since wascap works by using the Wasm binary as a Vec<u8> in memory and then adding the custom section.

The `embed_claims` function is the important part, specifically [this line](https://github.com/wasmCloud/wasmCloud/blob/main/crates/wascap/src/wasm.rs#L89) which uses the [wasm-gen](https://crates.io/crates/wasm-gen) crate to write the custom section.

So, all in all, to use this method you'd need the resulting binary as a slice, and you'd need to take the resulting Vec<u8> and do something productive with it to write the binary out

**Comment by Shnatsel** [comment]
The process should (hopefully) be as simple as writing the object file in the correct format, and then [telling the linker to also link it](https://github.com/rust-secure-code/cargo-auditable/blob/825fdce0f2904c9ee1177d4815439e0fc03b511c/cargo-auditable/src/rustc_wrapper.rs#L44-L50). No need to open the Pandora's box of linker scripts. 

We just need to figure out A) in what format our section needs to be written to be consumable by the linker and B) if the linker needs [some special arguments](https://github.com/rust-secure-code/cargo-auditable/blob/825fdce0f2904c9ee1177d4815439e0fc03b511c/cargo-auditable/src/rustc_wrapper.rs#L52-L59) to prevent it from removing the section, which [seems not to be the case](https://github.com/rust-secure-code/cargo-auditable/issues/53#issuecomment-2012492864).

**Comment by Shnatsel** [comment]
So I `strace`d the build process, and it seems the `wasm32-unknown-unknown` target uses `rust-lld` bundled with the toolchain (as opposed to the default system linker), at least on my Linux machine.

The trick is going to be writing the correct WASM(?) blob to pass to it. I've inspected the intermediate artifacts passed to the linker, and `file` command identifies them as "WebAssembly (wasm) binary module version 0x1 (MVP)". As long as we can write a file in that format with a single section with the audit data, we're golden.

If you can make a proof of concept that writes a standalone WASM file with a section with any contents, which when passed into `RUSTFLAGS=-Clink-arg=/path/to/your/file.wasm cargo build --release`  ends up in the final WASM binary, I can take it from here and fill in all the bits specific to `cargo auditable` myself.

**Comment by bjorn3** [comment]
Wasm is a very different format from any other object file formats. It for example requires function declarations and definitions to specify the function signature. It also has many constructs not representable by machine code like tables and (multiple) linear memories. As such adding wasm writing support to the object crate would likely be like trying to fit a square peg in a round hole. Using the wasm-encoder crate like rustc does now I believe would be a lot easier.

**Comment by Shnatsel** [comment]
I've got a proof of concept working using the `wasm-gen` crate. Please put this WASM blob through the wringer (wasm-opt, any other optimization and minification you run) and let me know if the`.dep-v0` custom section survives. You can verify that it survived using https://www.wa2.io

[wasm_play.wasm.gz](https://github.com/rust-secure-code/cargo-auditable/files/15189210/wasm_play.wasm.gz)

**Comment by brooksmtownsend** [comment]
In the textual format, not indexed by `.dep-v0` but possible that it's not an issue:
```
➜ cat auditable.wat|grep 'wasm-play'
  (data $.rodata (;0;) (i32.const 1048576) "Hello, wasm-play!")
```

Once optimized (same result for 3, 4, size, super-size passes with wasm-opt):
```
➜ ./bin/wasm-opt auditable.wasm -O4 -o four.wasm
➜ ./bin/wasm-opt auditable.wasm -O3 -o three.wasm
➜ ./bin/wasm-opt auditable.wasm -Os -o size.wasm
➜ ./bin/wasm-opt auditable.wasm -Oz -o supersize.wasm
➜ ls | grep "\.wasm" | xargs -I {} wasm-tools print {} |grep wasm-play
  (data (;0;) (i32.const 1048576) "Hello, wasm-play!")
  (data (;0;) (i32.const 1048576) "Hello, wasm-play!")
  (data (;0;) (i32.const 1048576) "Hello, wasm-play!")
  (data (;0;) (i32.const 1048576) "Hello, wasm-play!")
```

I also tried after enabling `--memory-packing` which should optimize data sections, same result.

Same thing as well with another custom section introduced (signing with a wasmCloud tool):

```
➜ wash claims sign default.wasm --name auditable --rev 0 --ver v0.1.0

Successfully signed ./default_s.wasm

➜ wash inspect ./default_s.wasm


                            auditable - Component
  Account         ABJ6WQ5FBGXECIJBBGA7BG73IKE7ITD2DRAURV5WB5ZUTYTNOOAYGPIG
  Component       MAVEPQ57VDYMDELOCTP4P4TX45VY3XQLGGVUVIHFFTYYC5IGR56M3CFI
  Expires                                                            never
  Can Be Used                                                  immediately
  Version                                                       v0.1.0 (0)
  Embedded WIT                                                       false
                                    Tags
  None

➜ wasm-tools print default_s.wasm | grep wasm-play
  (data (;0;) (i32.const 1048576) "Hello, wasm-play!")
```

One last thing you may want to try @Shnatsel is adding this to your `Cargo.toml` if you didn't have it already and compiling for release, to make sure stripping symbols doesn't affect this. Since my (wasmcloud's) tooling is after this step, I just want to be sure:
```toml
[profile.release]
# Optimize for small code size
lto = true
opt-level = "s"
strip = true
```

**Comment by Shnatsel** [comment]
```toml
[profile.release]
# Optimize for small code size
lto = true
opt-level = "s"
strip = true
```

I've verified that the custom section is preserved in this configuration. Its contents are not "Hello, wasm-play!" but "blah", so that's the string you should be grepping for.

**Comment by Shnatsel** [comment]
A branch with WASM support is available for testing: #145

**Comment by brooksmtownsend** [comment]
> Its contents are not "Hello, wasm-play!" but "blah", so that's the string you should be grepping for.

Whoops, my mistake there. https://github.com/xtuc/wasm-custom-section turned out to be a better debugging tool:
```
➜ wasm-custom-section ./auditable.wasm show .dep-v0
Section `.dep-v0` (4 bytes):
Length: 4 (0x4) bytes
0000:   62 6c 61 68                                          blah

➜ wasm-custom-section ./default.wasm show .dep-v0
Section `.dep-v0` (4 bytes):
Length: 4 (0x4) bytes
0000:   62 6c 61 68                                          blah

➜ wasm-custom-section ./default_s.wasm show .dep-v0
Section `.dep-v0` (4 bytes):
Length: 4 (0x4) bytes
0000:   62 6c 61 68                                          blah

➜ wasm-custom-section ./four.wasm show .dep-v0
Section `.dep-v0` (4 bytes):
Length: 4 (0x4) bytes
0000:   62 6c 61 68                                          blah
```

**Comment by Shnatsel** [comment]
Well, WASM support is basically complete in #145

Please try it and let me know what you think!

**Comment by Shnatsel** [comment]
I've published v0.6.3 with WebAssembly support.

I'll add WASM support to `cargo audit` shortly.

**Comment by Shnatsel** [comment]
PR for `cargo audit`: https://github.com/rustsec/rustsec/pull/1182
